# Linear Models

Linear model regression model is the first model we tried and will be baseline model

In [1]:
import utils
import gc
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression

## Read saved data sets

Read saved pickle format data sets from `1_Data_Exploration.ipynb`

In [2]:
X_train = pd.read_pickle("X_train.pkl")
X_val = pd.read_pickle("X_val.pkl")
X_test = pd.read_pickle("X_test.pkl")
y_train = pd.read_pickle("y_train.pkl")
y_val = pd.read_pickle("y_val.pkl")
y_test = pd.read_pickle("y_test.pkl")

## Implement linear regression 

* linear model with using `major` and without `major_new`

In [3]:
vars_drop = ['major_new']
X_train_1, X_val_1, X_test_1 = utils.drop_vars(vars_drop, X_train, X_val, X_test) 

In [4]:
X_train_1.head()

jobType       degree     major   industry  yearsExperience  \
68609   VICE_PRESIDENT      MASTERS      MATH  EDUCATION                5   
924598             CTO     DOCTORAL      MATH       AUTO               22   
918523             CEO  HIGH_SCHOOL      NONE  EDUCATION               24   
213733             CFO  HIGH_SCHOOL      NONE    FINANCE               22   
246703  VICE_PRESIDENT      MASTERS  BUSINESS    SERVICE               18   

        milesFromMetropolis  
68609                    82  
924598                   66  
918523                   67  
213733                   90  
246703                   68

In [5]:
var_cate = utils.get_categorical_variables(X_train_1)
X_train_hot_encode, X_valid_hot_encode = utils.encoding('one hot', var_cate, X_train_1, X_val_1)
_, X_test_hot_encode = utils.encoding('one hot', var_cate, X_train_1, X_test_1)

Check the multicollinearity before implemeting linear regression, numerical variables `yearsExperience` and `milesFromMetropolis` do not have multicollinearity due to their VIF less than 2. Categorical varibles are impossible to tell here.

In [6]:
utils.VIF(X_train_hot_encode)

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/statsmodels/stats/outliers_influence.py:193: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)


                   feature       VIF
0          yearsExperience  1.000034
1      milesFromMetropolis  1.000024
2              jobType_CEO       inf
3              jobType_CFO       inf
4              jobType_CTO       inf
5          jobType_JANITOR       inf
6           jobType_JUNIOR       inf
7          jobType_MANAGER       inf
8           jobType_SENIOR       inf
9   jobType_VICE_PRESIDENT       inf
10        degree_BACHELORS       inf
11         degree_DOCTORAL       inf
12      degree_HIGH_SCHOOL       inf
13          degree_MASTERS       inf
14             degree_NONE       inf
15           major_BIOLOGY       inf
16          major_BUSINESS       inf
17         major_CHEMISTRY       inf
18           major_COMPSCI       inf
19       major_ENGINEERING       inf
20        major_LITERATURE       inf
21              major_MATH       inf
22              major_NONE       inf
23           major_PHYSICS       inf
24           industry_AUTO       inf
25      industry_EDUCATION       inf
2

In [7]:
linear_reg = LinearRegression(normalize = True).fit(X_train_hot_encode, y_train)
linear_pred_train = linear_reg.predict(X_train_hot_encode)
linear_pred_val = linear_reg.predict(X_valid_hot_encode)
linear_pred_test = linear_reg.predict(X_test_hot_encode)

utils.check_RMSE("linear model train", y_train, linear_pred_train)
utils.check_RMSE("linear model valid", y_val, linear_pred_val)
utils.check_RMSE("linear model test", y_test, linear_pred_test)

RMSE of model linear model train is:  19.567400073310356
RMSE of model linear model valid is:  19.60858171041308
RMSE of model linear model test is:  19.5983063417872


* linear model without using `major`

In [8]:
vars_drop = ['major']
X_train_2, X_val_2, X_test_2 = utils.drop_vars(vars_drop, X_train, X_val, X_test) 
var_cate = utils.get_categorical_variables(X_train_2)
X_train_hot_encode, X_valid_hot_encode = utils.encoding('one hot', var_cate, X_train_2, X_val_2)
_, X_test_hot_encode = utils.encoding('one hot', var_cate, X_train_2, X_test_2)

Check the multicollinearity before implemeting linear regression, numerical variables `yearsExperience` and `milesFromMetropolis` do not have multicollinearity due to their VIF less than 2. Categorical varibles are impossible to tell here.

In [9]:
utils.VIF(X_train_hot_encode)

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/statsmodels/stats/outliers_influence.py:193: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)


                   feature       VIF
0          yearsExperience  1.000026
1      milesFromMetropolis  1.000021
2              jobType_CEO       inf
3              jobType_CFO       inf
4              jobType_CTO       inf
5          jobType_JANITOR       inf
6           jobType_JUNIOR       inf
7          jobType_MANAGER       inf
8           jobType_SENIOR       inf
9   jobType_VICE_PRESIDENT       inf
10        degree_BACHELORS       inf
11         degree_DOCTORAL       inf
12      degree_HIGH_SCHOOL       inf
13          degree_MASTERS       inf
14             degree_NONE       inf
15           industry_AUTO       inf
16      industry_EDUCATION       inf
17        industry_FINANCE       inf
18         industry_HEALTH       inf
19            industry_OIL       inf
20        industry_SERVICE       inf
21            industry_WEB       inf
22          major_new_NONE       inf
23      major_new_Not None       inf


In [10]:
linear_reg = LinearRegression(normalize = True).fit(X_train_hot_encode, y_train)
linear_pred_val = linear_reg.predict(X_valid_hot_encode)
linear_pred_test = linear_reg.predict(X_test_hot_encode)
model = "linear model"
utils.check_RMSE(model, y_val, linear_pred_val)
utils.check_RMSE(model, y_test, linear_pred_test)

RMSE of model linear model is:  19.807425626527287
RMSE of model linear model is:  19.803692476116765


## Summary

* The Linear regression model using `major` but `major_new` has RMSE 19.5983, which is the baseline performance.
* Do not use `major_new` because it will lose variable information and reduce model performance